src/mars.c

```diff
@@ -1658,35 +1659,50 @@ Language changes listed by -transition=id:\n\
         }
     }
 
-    if (!global.params.obj)
+    //printf("global.params.multiobj = %d, oneobj = %d, lib = %d\n",
+    //    global.params.multiobj, global.params.oneobj, global.params.lib);
+    if (!global.params.obj || !modules.dim)
     {
     }
     else if (global.params.oneobj)
     {
-        if (modules.dim)
-            obj_start(modules[0]->srcfile->toChars());
+        /* global.params.oneobj == true:
+         *  Just only one object file is generated for the final link.
+         *  e.g.
+         *      dmd -ofout main.d   // main.obj is generated
+         */
+        obj_start(modules[0]->srcfile->toChars());
+
         for (size_t i = 0; i < modules.dim; i++)
         {
             Module *m = modules[i];
             if (global.params.verbose)
                 fprintf(global.stdmsg, "code      %s\n", m->toChars());
+
             genObjFile(m, false);
             if (entrypoint && m == rootHasMain)
                 genObjFile(entrypoint, false);
         }
-        for (size_t i = 0; i < Module::amodules.dim; i++)
+        for (size_t j = 0; j < Module::amodules.dim; j++)
```

Comment:
In below code blocks, `j` is used to iterate non-root modules. It's just for the consistency.